framework/src/util/FrameworkFactory.cpp

```diff
@@ -28,10 +28,46 @@
 
 namespace cppmicroservices {
 
+// A helper class to shut down the framework when all external
+// references (Bundle, BundleContext, etc.) are destroyed. It
+// also manages the lifetime of the CoreBundleContext instance.
+struct CoreBundleContextHolder
+{
+  CoreBundleContextHolder(std::unique_ptr<CoreBundleContext> ctx)
+    : ctx(std::move(ctx))
+  {}
+
+  ~CoreBundleContextHolder()
+  {
+    auto const state = ctx->systemBundle->state.load();
+    if (((Bundle::STATE_STARTING | Bundle::STATE_ACTIVE) & state) == 0)
+    {
+      // Call WaitForStop in case some did call Framework::Stop()
+      // but didn't wait for it. This joins with a potentially
+      // running framework shut down thread.
+      ctx->systemBundle->WaitForStop(std::chrono::milliseconds::zero());
+      return;
+    }
+
+    // Create a new CoreBundleContext holder, in case some event listener
```

Comment:
The added test case already relies on that code path. There is no standard way to query the `Bundle` object if it internally holds a "new" `CoreBundleContextHolder` though. The test assumes that this just "works", otherwise the OS should throw segmentation faults (which it did) when we try to access an invalid pointer to the `CoreBundleContext` in the test.